src/common/Gateway.sol

```diff
@@ -308,6 +308,7 @@ contract Gateway is Auth, IGateway, Recoverable {
             }
         }
 
+        transactionPayer = address(0);
```

Comment:
@lemunozm previously, a smart contract (including our tests) doing
```solidity
poolManager.registerAsset{value: EXACT_GAS}(..);
someVault.requestDeposit(..);
```

would fail since the `requestDeposit` call would not work subsidized as `transactionPayer` was still set.

Reply:
The issue is the test is missing /// forge-config: default.isolate = true

Reply:
Note that here, you're resetting the transaction payer, but you could still be in a batch where more messages needs to be sent. i.e:

- FM.multicall{PAY}()
  - NotifyPoolA // Will be paid with PAY
  - NotifyPoolB // Will be subsidized when it shouldn't

Reply:
> CR: The issue I think is the test is missing /// forge-config: default.isolate = true

That didn't fix it. I believe because `registerAsset` is called in the setup of the pool, which is part of the test.

> Note that here, you're resetting the transaction payer, but you could still be in a batch where more messages needs to be sent. i.e:

I don't get this, wouldn't `send()` never be triggered until the end of the `multicall`?

Reply:
Gotcha! Should we add a `resetTransactionPayer()` method that is called at the end of `PoolManager.registerAsset/transerShares`?

Reply:
I think this would actually be valuable though, not just for tests.

You might very well have a smart contract integration that wants to do
```solidity
poolManager.registerAsset{value: EXACT_GAS}(..);
someVault.requestDeposit(..);
```

I believe that should be possible. Without some fix for resetting the payment mechanism, the above will never work.

Reply:
Especially with the upcoming wider adoption of smart contract wallets.

Reply:
Ok, then, let's add it!

Just concerned about the security issues. Anyone could call to it in the middle of a transaction to disable the payment and use the subsidized pool and this kind of things.

I think it should be `auth`, and we should call it after all methods where we already call `_pay()`, probably it should be a modifier as follows:

```
modifier transactionPayment() {
    if (!gateway.isBatching())) gateway.payTransaction(msg.sender);
    _;
    if (!gateway.isBatching())) gateway.resetTransactionPayer();
}
```

WDYT?

Something similar we would need to do in the multicall

Reply:
Yes it should definitely be `auth`

Reply:
I don't think this is needed on the `Hub`, since there is no subsidized transactions possible on the Hub side.

Reply:
> since there is no subsidized transactions possible on the Hub side

Just to note (although with the latest solution I think it does not apply). There are subsidized messages in some cases on the Hub side. When we a CV message reach CP and dispatch another message to CV, that message is subsidized.

Reply:
Good point!